### DISCUSSION TITLE
How to poll the current image processing progress?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5868] [Development] [answered]
I am aware about methods `MagickSetProgressMonitor` and `MagickSetImageProgressMonitor`, but these methods register callbacks to be called when the progress changes (i.e. after `MagickSetProgressMonitor` or `MagickSetImageProgressMonitor` returns), and currently such methods can not be called from dart language through dart ffi (I am creating bindings for IM in dart).
I saw a method called `MonitorProgress` but It looks to be private method as it is decalred `static`.
So is there equivalent methods to these methods which lets me get the current progress synchronously, maybe something like `MagickGetProgress` or `MagickGetImageProgress`? Is there such methods?

**Comment by urban-warrior** [comment, marked as answer]
ImageMagick does not maintain state in regards to progress.  Instead, progress state can only be captured by the monitor callback.  ImageMagick includes a default callback, otherwise the user must specify a custom callback and only then can the user capture state information which includes a message, a progress offset, and the maximum extent of the offset.  We could of course add image artifacts that create a key value pair to track progress, e.g. `progress:offset` obtained with the GetImageArtifact() method.  Let us know if that would resolve your issue, and we'll commit a patch.

**Reply by Haidar0096** [reply]
@urban-warrior thx for your reply.

having an option for a method that returns the progress and terminates instead of only having the option to register a callback will be much better, I would appreciate an update that adds such feature.

so if I understand you correctly you will save the latest progress as an artifact on the image each time the progress is updated? and that progress can be polled by the key "progress"? If the answer is yes then I think you may also want to set other key:value pairs as image artifacts because the registered callback (monitor function) is called from various events in the magick core api, so you maybe would set a key:value pair for each type of progress (ex: offset:offsetvalue tag:tagvalue ...)

**Reply by urban-warrior** [reply]
Because the current progress monitoring is stateless, a method to return the progress is not possible-- unless we start tracking state.  This is exactly why the callback is needed, so the consumer can track state or terminate if they want.    Saving state as image artifacts is simple enough, we just save the tag, offset, and extent, and the consumer can query with the existing GetImageArtifact() method.  No state  info is needed.  However, there is no way to terminate.   We could reserve a namespace to terminate-- here the consumer sets, for example, `monitor:terminate=true` and if monitoring is enabled, ImageMagick would unwind and exit.  If we return to your proposal, we would need to add state so the new method can return the state and possibly terminate.  This is a lot of overhead when the existing callback is specifically designed for this use case.

You are essentially asking for a backdoor to a well-defined process simply because DART:FFI does not have a monitor callback?  Could you not add one?

Review the above and propose a workable solution and we will consider implementing it.  Thus far, we're trending toward leaving progress monitoring as is.

**Reply by Haidar0096** [reply]
@urban-warrior 
There is always a workaround, I already found a workaround in dart, but I asked here to make sure there is no option I have missed.

I understand that adding state is a big overhead, let alone adding state with termination. So I will try the workaround I found and see how it goes, and if you decided to save the state as an image artifact, please let me know to add this as a binding as well.

Thank you for your responses and great library :)

**Reply by urban-warrior** [reply]
We'll add an image artifact.  It will only be saved if monitoring is enabled.  Use GetImageArtifact() to retrieve it with the `monitor:progress` as the key.  The value format is `% complete:tag:filename`, e.g. `80.0%:Morphology/Image:rose.png`.     If you have a better format for the value, let us know-- otherwise, look for the patch by sometime tomorrow.
